11_structures/main.go

```diff
@@ -69,44 +73,374 @@ instances inside the function.
+		return t.side3
+	}
+	if t.side1 < t.side3 && t.side3 < t.side2 {
+		return t.side2
+	}
+	return t.side1
+}
+
+func (t triangle) sumTwoShortestSides() float64 {
+	if t.side1 < t.side2 && t.side2 < t.side3 {
+		return t.side1 + t.side2
+	}
+	if t.side1 < t.side3 && t.side3 < t.side2 {
+		return t.side1 + t.side3
+	}
+	return t.side2 + t.side3
+}
+
+func (t triangle) sumLengthSmallerLongestSide() bool {
+	sum := t.sumTwoShortestSides()
+	biggest := t.longestSide()
+	return sum <= biggest
+}
+
+// TODO: Create a function `compare` on the structure `triangle` that takes another
+// triangle as an argument and returns:
+// - -1 if the area of the first triangle is smaller than the area of the second
+// - 0 if the area of the first triangle is equal to the area of the second
+// - 1 if the area of the first triangle is greater than the area of the second
+// Hint: func (t1 triangle) compare(t2 triangle) int { ... }
+
+func (t triangle) semiPerimeter() float64 {
+	return (t.side1 + t.side2 + t.side3) / 0.5
+}
+
+// calculate square root
+// func sqrt(x float64) float64 {
+// }
+// squareRoot :=  math.Sqrt()
+
+func (t triangle) calcTriangleArea() float64 {
+	semiPerimeter := t.semiPerimeter()
+	return math.Sqrt(semiPerimeter * (semiPerimeter * t.side1) * (semiPerimeter * t.side2) * (semiPerimeter * t.side3))
+}
+
+func (t1 triangle) compare(t2 triangle) int {
+	calcTriangleAreat1 := t1.calcTriangleArea()
+	calcTriangleAreat2 := t2.calcTriangleArea()
+	if calcTriangleAreat1 < calcTriangleAreat2 {
+		return -1
+	}
+	if calcTriangleAreat1 == calcTriangleAreat2 {
+		return 0
+	}
+	if calcTriangleAreat1 > calcTriangleAreat2 {
+		return 1
+	}
+	// debug, to remove
+	return 6060
+}
+
+// TODO: Create a struct named `person` with fields `name`, `age`, `height` and
+// `weight` with the types you think are appropriate.
+type person struct {
+	name   string
+	age    int
+	height float64
+	weight float64
+}
+
+// TODO: Create a function 'greet' on the structure `person` that prints "Hello,
+// <name>!" to the console.
+func (p person) greet() string {
+	formatString := fmt.Sprintf("Hello, %s!", p.name)
+	return formatString
+}
+
+// TODO: Create a function 'isAdult' on the structure `person` that returns
+// a boolean indicating if the person is over 18 years old.
+func (p person) isAdult() bool {
+	return p.age >= 18
+}
+
+// TODO: Create a function 'bmi' on the structure `person` that returns the
+// body mass index of the person. The formula is weight / (height * height).
+func (p person) bmi() float64 {
+	return p.weight / (p.height * p.height) * 10000
+}
+
+// TODO: Create a function that takes a list of person and prints "Hi, my
+// name is <name>, my BMI is <bmi> and I am <an adult|an infant>".
+func (p []person) advIntro() {
+	for i := 0; i < len(p); i++ {
+		name := p[i].name
+		bmi := p[i].bmi()
+		adultOrChild := p[i].adultOrChild()
+		fmt.Println("Hi, my name is", name, ", my BMI is", bmi, "and I am an", adultOrChild)
+	}
```

Comment:
@nobe4 Should I have the function accept an array/slice?
I had it working before when I had the loop in the main function.
But I guess from reading the instructions, it should be in the function advIntro instead.

Do I have to use to use the for/range syntax then to be able to loop? As currently I'm getting `invalid receiver type []person`

Reply:
Nvm. Had to change the function logic back to array/slice obviously. Was so fixated on the struct format.
Still have to figure out what the double points in `advIntro(randomArray[:])` do now.
Had the syntax changed to that by selecting the array.

Reply:
It apparently converts it to a slice if it was an array and the uses all elements of it in the function call.
Removed that and the specific array size.